src/Store.ts

```diff
@@ -0,0 +1,581 @@
+import * as arrayUtil from 'dojo-core/array';
+import * as aspect from 'dojo-core/aspect';
+import Evented from 'dojo-core/Evented';
+import { default as has, add as hasAdd } from 'dojo-core/has';
+import { EventObject, Handle, Hash } from 'dojo-core/interfaces';
+import * as lang from 'dojo-core/lang';
+import Promise from 'dojo-core/Promise';
+import Filter from './Filter';
+import * as dstore from './interfaces';
+import QueryMethod, { QueryMethodArgs } from './QueryMethod';
+
+// module:
+//		dstore/Store
+/* jshint proto: true */
+// detect __proto__, and avoid using it on Firefox, as they warn about
+// deoptimizations. The watch method is a clear indicator of the Firefox
+// JS engine.
+
+hasAdd('object-proto', Boolean((<any> {})['__proto__']) && !(<any> {}).watch);
+const hasProto = has('object-proto');
+
+export interface NewableStoreModel {
+	new (args?: {}): any;
+	createSubclass?: (arg: any[]) => any;
+	prototype: any;
+}
+
+export interface StoreArgs {
+	Model?: NewableStoreModel;
+}
+
+export default class Store<T> extends Evented implements dstore.Collection<T>, Hash<any> {
+	[ index: string ]: any;
+	// summary:
+	//		Creates an object, throws an error if the object already exists
+	// object: Object
+	//		The object to store.
+	// directives: dstore/Store.PutDirectives?
+	//		Additional directives for creating objects.
+	// returns: Object
+	//		The object that was stored, with any changes that were made by
+	//		the storage system (like generated id)
+	add: (object: any, directives?: dstore.PutDirectives) => any;
+
+	// autoEmitEvents: Boolean
+	//		Indicates if the events should automatically be fired for put, add, remove
+	//		method calls. Stores may wish to explicitly fire events, to control when
+	//		and which event is fired.
+	autoEmitEvents: boolean;
+
+	// autoEmitHandles: Handle[]
+	//		Holdes handles to the aspects responsible for automatically emitting events
+	//		when add, remove, or put are called.
+	// TODO - Once all classes are converted to typescript, and overriding value for autoEmitEvents
+	// can be passed to the constructor, and this will no longer be necessary.
+	autoEmitHandles: Handle[];
+
+	fetch: (args?: dstore.FetchArgs) => dstore.FetchPromise<T>;
+
+	Filter: typeof Filter;
+
+	// summary:
+	//		Retrieves an object by its identity
+	// id: Number
+	//		The identity to use to lookup the object
+	// returns: Object
+	//		The object in the store that matches the given id.
+	get: (id: string | number) => any;
```

Comment:
This should probably return `Promise<T> | void`.  (I only include `void` because I think Memory returnins undefined for nonexistent IDs.  We might want to make that more consistent eventually, I'll make a note.)